experimental/ixdtf/src/parser.rs

```diff
@@ -0,0 +1,400 @@
+                }
+            }
+            _ => match s[idx..(idx + 4)].parse::<i32>() {
+                Ok(year) => ((idx + 4), Some(year)),
+                Err(_) => (idx, None)
+            }
+        }
+    }
+
+    fn parse_date_month(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(month) => {
+                if month > 0 && month <= 12 {
+                    ((idx + 2), Some(month))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_date_day(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(day) => {
+                if day > 0 && day <= 31 {
+                    ((idx + 2), Some(day))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_time_hour(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(hour) => {
+                if hour <= 24 {
+                    ((idx + 2), Some(hour))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_time_minute(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(minute) => {
+                if minute < 60 {
+                    ((idx + 2), Some(minute))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_time_second(idx :usize, s: &str) -> (usize, Option<u8>) {
+        match s[idx..(idx + 2)].parse::<u8>() {
+            Ok(second) => {
+                if second <= 60 {
+                    ((idx + 2), Some(second))
+                } else {
+                    (0, None)
+                }
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+    fn parse_time_fractional_part(idx :usize, s: &str) -> (usize, Option<i32>) {
+        match s[idx..].parse::<i32>() {
+            Ok(fraction) => {
+                (s.len(), Some(fraction))
+            },
+            Err(_) => (0, None)
+        }
+    }
+
+
+    pub fn parse(s: String) -> Self {
+        let len = s.len();
+        let mut year : Option<i32> = None;
+        let mut month : Option<u8> = None;
+        let mut day : Option<u8> = None;
+        let mut hour : Option<u8> = None;
+        let mut minute : Option<u8> = None;
+        let mut second : Option<u8> = None;
+        let mut nano_second : Option<i32> = None;
+        let mut idx : usize = 0;
+
+        (idx, year) = Self::parse_date_year(idx, &s);
+        if (idx == 0) {
+            return Self::empty()
+        }
+        if idx < len && &s[idx..(idx+1)] == "-" {
```

Comment:
Suggestion: Indexing operations in Rust can panic. It would be easier to reason about the code not panicking if you used helper functions like [split_first](https://doc.rust-lang.org/std/primitive.slice.html#method.split_first) or if you converted the slice to an iterator and pulled items off with `.peek()` and `.next()`. If you must use indexing operators, make sure they are guarded by a precondition so that they don't panic.

It will be hard to write this code so that it never panics, but this is a property we definitely want to have.